rosbag2_cpp/src/rosbag2_cpp/cache/message_cache.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020, Robotec.ai sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "rosbag2_cpp/cache/message_cache.hpp"
+#include "rosbag2_cpp/logging.hpp"
+
+namespace rosbag2_cpp
+{
+namespace cache
+{
+
+MessageCache::MessageCache(const uint64_t & max_buffer_size)
+{
+  primary_buffer_ = std::make_shared<MessageCacheBuffer>(max_buffer_size);
+  secondary_buffer_ = std::make_shared<MessageCacheBuffer>(max_buffer_size);
+}
+
+bool MessageCache::push(std::shared_ptr<const rosbag2_storage::SerializedBagMessage> msg)
+{
+  // While pushing, we keep track of inserted and dropped messages as well
+  bool pushed = false;
+  {
+    std::lock_guard<std::mutex> buffer_lock(buffer_mutex_);
+    pushed = primary_buffer_->push(msg);
+  }
+
+  allow_swap();
```

Comment:
Can elaborate this for me once more? If I allow for disc-write the buffer after each push to the buffer, how can be buffer ever be full? Shouldn't we allow the swap only when `primary_buffer_->push(msg)` returns `false`, indicating that the primary buffer is full?

Reply:
The key here is the **max**_cache_size 
How can buffer ever be full?
- The consumer thread always consumes the whole consumer buffer, which can have 1 or more messages
- Consuming involves storage writing, which could take considerable time (and the time can spike)
- During this time, producer buffer can fill to some degree (up to max cache)
- Cache being full is a kind of pathology here - since we start dropping messages right when this happens
This implementation is centered around optimizing I/O - we want to guarantee that consumer thread is always busy when there is any amount of data to be written to disc. This might not be very important in average case, but when the throughput is near the disk write limit, this approach makes a difference.

Reply:
> Consuming involves storage writing, which could take considerable time (and the time can spike)

I think this is the key point - while a write is going on that's when we're filling up the other buffer - the more traffic there is, the more will go into the buffer and the longer the write will take, so high throughput scenarios actually have less frequent but much bigger write transactions

Reply:
understood. I would love to see a similar text as in your answer being part of the documentation for these cache classes. I think that makes a lot of sense to document the intend here.